pkg/controller/infinispan/infinispan_controller.go

```diff
@@ -239,56 +258,78 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
 			return reconcile.Result{}, err
 		}
 
-		ser := r.serviceForInfinispan(infinispan)
-
-		if err != nil {
-			reqLogger.Error(err, "failed to update Infinispan")
-			return reconcile.Result{}, err
-		}
+		// StatefulSet created successfully
+		reqLogger.Info("End of the StetefulSet creation")
+	}
+	if err != nil {
+		reqLogger.Error(err, "failed to get StatefulSet")
+		return reconcile.Result{}, err
+	}
 
-		setupServiceForEncryption(infinispan, ser)
-		err = r.client.Create(context.TODO(), ser)
-		if err != nil && !errors.IsAlreadyExists(err) {
-			reqLogger.Error(err, "failed to create Service", "Service", ser)
-			return reconcile.Result{}, err
-		}
+	ser := r.computeService(infinispan)
+	setupServiceForEncryption(infinispan, ser)
 
-		serDNS := r.serviceForDNSPing(infinispan)
-		err = r.client.Create(context.TODO(), serDNS)
-		if err != nil && !errors.IsAlreadyExists(err) {
-			reqLogger.Error(err, "failed to create Service", "Service", serDNS)
-			return reconcile.Result{}, err
-		}
-		reqLogger.Info("Created Service", "Service", serDNS)
+	err = r.reconcileService(infinispan, ser, reqLogger)
+	if err != nil {
+		reqLogger.Error(err, "Error in reconcileService", "Infinispan.Namespace", infinispan.Namespace)
+		return reconcile.Result{}, err
+	}
 
-		// StatefulSet created successfully - return and requeue
-		reqLogger.Info("End of the StetefulSet creation")
-		return reconcile.Result{Requeue: true}, nil
+	serDNS := r.computePingService(infinispan)
 
-	} else if err != nil {
-		reqLogger.Error(err, "failed to get StatefulSet")
+	err = r.reconcileService(infinispan, serDNS, reqLogger)
+	if err != nil {
+		reqLogger.Error(err, "Error in reconcileService for DNS", "Infinispan.Namespace", infinispan.Namespace)
 		return reconcile.Result{}, err
-	} else if infinispan.IsUpgradeNeeded(reqLogger) {
+	}
+
+	if infinispan.IsExposed() {
+		switch infinispan.Spec.Expose.Type {
+		case infinispanv1.ExposeTypeLoadBalancer, infinispanv1.ExposeTypeNodePort:
+			externalService := r.computeServiceExternal(infinispan)
+			result, err := r.reconcileExternalService(infinispan, externalService, reqLogger)
+			if result != nil {
+				return *result, err
+			}
+		case infinispanv1.ExposeTypeRoute:
+			ok, err := r.isGroupVersionSupported(routev1.SchemeGroupVersion.String(), "Route")
+			if err != nil {
+				reqLogger.Error(err, fmt.Sprintf("Failed to check if %s is supported", routev1.SchemeGroupVersion.String()))
+				// Log an error and try to go on with Ingress
+				ok = false
+			}
+			if ok {
+				route := r.computeRoute(infinispan, reqLogger)
+				result, err := r.reconcileRoute(infinispan, route, reqLogger)
+				if result != nil {
+					return *result, err
+				}
+			} else {
+				ingress := r.computeIngress(infinispan, reqLogger)
+				result, err := r.reconcileIngress(infinispan, ingress, reqLogger)
+				if result != nil {
+					return *result, err
+				}
+			}
+		default:
```

Comment:
Looks like we can free to remove default section as now Expose.Type validates before Reconcile loop inside the enums